### PR TITLE
slow tpu train

### LINKED ISSUE #2016: specifying the tpu_core speed-up TPU training
## 🐛 Bug
1. I am getting a huge time difference between training a model on a specific tpu core tpu_cores=[1] and training a model on just 1 tpu core tpu_cores=1. What's the reason for that? Aren't both the conditions the same with just the difference that I am assigning a specific tpu_core in the first case and assigning the number of tpu_cores I want to use in the second case. Also in the second case, I am getting an error. When training with tpu_cores=[1] epoch time is 17 seconds with tpu_cores=1 epoch time is just 5 seconds.
2. Running on colab gives me an error but no error on Kaggle kernels. But the time difference issue is the same on both the platforms.

#### Code sample
[Colab Notebook](https://colab.research.google.com/drive/1jPen-P-e4njk_vHUQPsNWrcJ87P5UFxr?usp=sharing)

### Expected behavior

As far as I know in both cases, the training time should be the same regardless of training on a single core or training on a specific core.

### Environment
 - PyTorch Version (e.g., 1.0): 1.5.0
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.7
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: Tesla P100-PCIE-16GB
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by lezwon**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## What does this PR do?
Fixes #2016 .

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -412,8 +412,8 @@ def run_training_epoch(self):
         train_dataloader = self.train_dataloader
 
         # on TPU we have to wrap it under the ParallelLoader
-        if self.use_tpu and self.tpu_id is None:
-            device = xm.xla_device()
+        if self.use_tpu:
+            device = xm.xla_device(self.tpu_id)
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py:416]: what if `self.tpu_id` is None?

**Comment by rohitgr7** [comment]
Working fine on single tpu_core and specific tpu_core but now getting an error `RuntimeError: Cannot replicate if number of devices (1) is different from 8` when training on all the 8 cores `tpu_cores=8`. [This might be the reason](https://github.com/pytorch/xla/issues/1576#issuecomment-580978427).

**Comment by williamFalcon** [comment]
once this is merged:
https://github.com/PyTorchLightning/pytorch-lightning/pull/2029

Let's get rid of .spawn and replace with the same method

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (lezwon) [reply pytorch_lightning/trainer/training_loop.py]: Sorry. Missed that out. Have added a condition. :]

**Comment by lezwon** [comment]
> Working fine on single tpu_core and specific tpu_core but now getting an error `RuntimeError: Cannot replicate if number of devices (1) is different from 8` when training on all the 8 cores `tpu_cores=8`. [This might be the reason](https://github.com/pytorch/xla/issues/1576#issuecomment-580978427).

@rohitgr7 This usually happens when you have already selected a core i.e `tpu_cores=[1]` and later try to train on all cores `tpu_cores=8`. If you select `tpu_cores=8` on the first attempt, it should work fine. :]

**Comment by Borda** [comment]
@lezwon mind add note to changelog?

File: pytorch_lightning/trainer/evaluation_loop.py [hunk 2]
```diff
@@ -249,8 +249,8 @@ def _evaluate(self, model: LightningModule, dataloaders, max_batches: int, test_
             dl_outputs = []
 
             # on TPU we have to wrap it under the ParallelLoader
-            if self.use_tpu and self.tpu_id is None:
-                device = xm.xla_device()
+            if self.use_tpu:
+                device = xm.xla_device(self.tpu_id) if self.tpu_id is not None else xm.xla_device()
```
Comment (awaelchli) [inline pytorch_lightning/trainer/evaluation_loop.py:253]: According to docs, the None case does not need to be handled explicitly:
https://pytorch.org/xla/release/1.5/index.html#torch_xla.core.xla_model.xla_device
`device = xm.xla_device(self.tpu_id)` is fine even if `self.tpu_id = None`

Comment (lezwon) [reply pytorch_lightning/trainer/evaluation_loop.py:253]: @awaelchli Yea, it does work without it too. We could refactor these instances across the project.

Comment (awaelchli) [reply pytorch_lightning/trainer/evaluation_loop.py:253]: I have refactored it here #1756 so if you do it here too, then I think it's pretty much all.

Comment (Borda) [reply pytorch_lightning/trainer/evaluation_loop.py:253]: Does it also work for older versions... =)

**Comment by rohitgr7** [comment]
@lezwon I ran directly with 8 cores and that error occured. I see you have made some changes in the PR. Will try again with updated version tonight and let you know.

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (williamFalcon) [reply pytorch_lightning/trainer/training_loop.py:416]: where is the condition? i still don't see where you check for None

File: pytorch_lightning/trainer/evaluation_loop.py (hunk 2 above)
Comment (lezwon) [reply pytorch_lightning/trainer/evaluation_loop.py]: @Borda Do you mean older versions of lightning? I've noticed `xm.xla_device(None)` working earlier too. The docs @awaelchli pointed out though, confirms that it should work well.

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/trainer/training_loop.py:416]: as mentioned in concurrent PR, `None` is allowed option

File: pytorch_lightning/trainer/evaluation_loop.py (hunk 2 above)
Comment (Borda) [reply pytorch_lightning/trainer/evaluation_loop.py]: I was thinking of older version of XLA but there is no reason why it should not :]

**Comment by rohitgr7** [comment]
Getting this error now in all the cases, even when is set `checkpoint_callback=False`
```
KeyError                                  Traceback (most recent call last)
<ipython-input-17-31113c96baf3> in <module>
      1 # Train on specific tpu_core
----> 2 train_validate(0, [7])

<ipython-input-16-7fc43d3b07cf> in train_validate(fold_i, tpu_cores)
     21 #         fast_dev_run=True
     22     )
---> 23     trainer.fit(model, train_dl, valid_dl)
     24     torch.cuda.empty_cache()

/opt/conda/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py in fit(self, model, train_dataloader, val_dataloaders)
    910 
    911             # load weights if not interrupted
--> 912             self.load_spawn_weights(model)
    913             self.model = model
    914 

/opt/conda/lib/python3.7/site-packages/pytorch_lightning/trainer/distrib_data_parallel.py in load_spawn_weights(self, original_model)
    415             # load weights saved in ddp
    416             path = os.path.join(self.default_root_dir, '__temp_weight_ddp_end.ckpt')
--> 417             loaded_model = original_model.__class__.load_from_checkpoint(path)
    418 
    419             # copy loaded weights to old model

/opt/conda/lib/python3.7/site-packages/pytorch_lightning/core/lightning.py in load_from_checkpoint(cls, checkpoint_path, map_location, hparams_file, tags_csv, *args, **kwargs)
   1561 
   1562         # override the module_arguments with values that were passed in
-> 1563         checkpoint[CHECKPOINT_KEY_MODULE_ARGS].update(kwargs)
   1564 
   1565         model = cls._load_model_state(checkpoint, *args, **kwargs)

KeyError: 'module_arguments'
```

**Comment by awaelchli** [comment]
the checkpoints trained with an old version (before hparams PR) are not compatible with the latest change to how hparams are stored in the checkpoints. This is a pending issue that needs to be resolved. 
You would have to retrain the model or go and manually modify the checkpoint to set the module_arguments key.

**Comment by rohitgr7** [comment]
I retrained the model from scratch.

**Comment by pvnieo** [comment]
Hi @rohitgr7 @lezwon @awaelchli 
I updated my pytorch-lightning package to the version available yesterday in master, But i just noticed that the `hparams.yaml` file is empty contrary to what was the default behavior before, also, when I wanted to test my model, I got the following error:
```
File "/home/.local/lib/python3.8/site-packages/pytorch_lightning/core/lightning.py", line 1563, in load_from_checkpoint
    checkpoint[CHECKPOINT_KEY_MODULE_ARGS].update(kwargs)
KeyError: 'module_arguments'
```
Is this issue solved yet?

**Comment by williamFalcon** [comment]
@Borda FYI

we’re, working on fixing the hparams thing

File: CHANGELOG.md [hunk 3]
```diff
@@ -20,6 +20,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Attribute `best_model_path` to `ModelCheckpoint` for storing and later retrieving the path to the best saved model file ([#1799](https://github.com/PyTorchLightning/pytorch-lightning/pull/1799)) 
 
+- Speed up single core TPU training by loading data using `ParallelLoader`.  ([#2033](https://github.com/PyTorchLightning/pytorch-lightning/pull/2033))
```
Comment (Borda) [inline CHANGELOG.md:23]:
```suggestion
- Speed up single-core TPU training by loading data using `ParallelLoader` ([#2033](https://github.com/PyTorchLightning/pytorch-lightning/pull/2033))
```

**Comment by pvnieo** [comment]
Hi @Borda @williamFalcon 
Is the issue concerning hparams not being saved, and the issue that arises when testing (KeyError: 'module_arguments') is solved?

**Pull request merged**